### PR TITLE
Prompt caching

### PR DESCRIPTION
**Pull request by sani903**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**
Added prompt caching for Anthropic models along with test cases to verify if caching is working.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**
https://github.com/OpenDevin/OpenDevin/issues/3397
1. Detects Anthropic model and caches the latest prompts (greater than 1024 tokens). 
2. Processes prompts to other models separately
Anthropic does not allow caching of message history shorter than 1024 tokens, but instead of throwing a BadRequestError if cache-control is used for the message, it now processes shorter messages without caching them. Thus, caching can be on by default.
Can add other models as they allow caching.

---
**Other references**

### CODE REVIEW

File: opendevin/llm/llm.py [hunk 1]
```diff
@@ -178,7 +177,31 @@ def wrapper(*args, **kwargs):
...
+                            message['content'], list
+                        ):
+                            for content in message['content']:
+                                if isinstance(content, dict) and 'text' in content:
+                                    # Add cache control to the content
+                                    content['cache_control'] = {'type': 'ephemeral'}
+
+                    # Set extra headers for Anthropic models
+                    extra_headers = {
+                        'anthropic-version': '2023-06-01',
```
Comment (enyst) [inline opendevin/llm/llm.py:196]:
```suggestion
```
This header doesn't seem necessary, liteLLM must be adding an api version by default. Please see also [discussion](https://github.com/OpenDevin/OpenDevin/pull/3411#discussion_r1719209811).

File: opendevin/llm/llm.py [hunk 2]
```diff
@@ -178,7 +177,31 @@ def wrapper(*args, **kwargs):
...
+                if is_anthropic_model:
+                    # Add cache control to the messages
+                    for message in messages:
+                        if 'content' in message and isinstance(
+                            message['content'], list
+                        ):
+                            for content in message['content']:
+                                if isinstance(content, dict) and 'text' in content:
+                                    # Add cache control to the content
+                                    content['cache_control'] = {'type': 'ephemeral'}
```
Comment (enyst) [inline opendevin/llm/llm.py:192]: There is a limit of 4 checkpoints, from Anthropic docs. Adding it to all messages doesn't seem to work.

**Comment by enyst** [comment]
Thanks for doing this! Please see also https://github.com/OpenDevin/OpenDevin/pull/3411

File: tests/unit/test_caching.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+import pytest
+
+from opendevin.core.config import LLMConfig
+from opendevin.llm.llm import LLM
+
+sample_text = """
```
Comment (enyst) [inline tests/unit/test_caching.py:6]: I love to see tests here, thanks!

File: opendevin/llm/llm.py (hunk 1 above)
Comment (sani903) [reply opendevin/llm/llm.py:196]: Okay, got it! Indeed without the header it is working the same

File: tests/unit/test_caching.py (hunk 3 above)
Comment (enyst) [inline tests/unit/test_caching.py:6]: I love that you set up tests! But wait, we don't run this in CI on Anthropic. Maybe we can mock here just to see that we send the right things, and for the rest it'll be something for our local machines? Cc: @li-boxuan

File: opendevin/llm/llm.py (hunk 2 above)
Comment (sani903) [reply opendevin/llm/llm.py:192]: Okay will work on this!

**Comment by li-boxuan** [comment]
> Thanks for doing this! Please see also #3411

It seems these two PRs are addressing the same task and #3411 came first?

File: tests/unit/test_caching.py (hunk 3 above)
Comment (li-boxuan) [reply tests/unit/test_caching.py:6]: I'd say the best we could do is to have some unittests checking the headers.

**Comment by enyst** [comment]
> > Thanks for doing this! Please see also #3411
> 
> It seems these two PRs are addressing the same task and #3411 came first?

Yes, and 3411 is almost ready to merge.

@sani903  please feel free to review the other PR or see if you wish to add something, it might be best so you don't have to duplicate work. Sorry this prompt caching issue is popular right now. 😅 But thank you for your contribution!

**Comment by sani903** [comment]
> > > Thanks for doing this! Please see also #3411
> > 
> > 
> > It seems these two PRs are addressing the same task and #3411 came first?
> 
> Yes, and 3411 is almost ready to merge.
> 
> @sani903 please feel free to review the other PR or see if you wish to add something, it might be best so you don't have to duplicate work. Sorry this prompt caching issue is popular right now. 😅 But thank you for your contribution!

Yes, got it. 3411 is at a pretty advanced stage and I have been following it as well. Will look to contribute on another issue then!

**Pull request closed without merging**